src/ReservoirRouter.sol

```diff
@@ -0,0 +1,305 @@
+        if (aCurveId == 0) {
+            rAmountOut = ReservoirLibrary.getAmountOutConstantProduct(aAmountIn, aReserveIn, aReserveOut, aSwapFee);
+        }
+        else if (aCurveId == 1) {
+            rAmountOut = ReservoirLibrary.getAmountOutStable(aAmountIn, aReserveIn, aReserveOut, aSwapFee, aExtraData);
+        }
+    }
+
+    function getAmountIn(
+        uint256 aAmountOut,
+        uint256 aReserveIn,
+        uint256 aReserveOut,
+        uint256 aCurveId,
+        uint256 aSwapFee,
+        ExtraData calldata aExtraData
+    ) external pure returns (uint256 rAmountIn) {
+        if (aCurveId == 0) {
+            rAmountIn = ReservoirLibrary.getAmountInConstantProduct(aAmountOut, aReserveIn, aReserveOut, aSwapFee);
+        }
+        else if (aCurveId == 1) {
+            rAmountIn = ReservoirLibrary.getAmountInStable(aAmountOut, aReserveIn, aReserveOut, aSwapFee, aExtraData);
+        }
+    }
+
+    function getAmountsOut(
+        uint256 aAmountIn,
+        address[] calldata aPath,
+        uint256[] calldata aCurveIds
+    ) external view returns(uint256[] memory rAmountsOut) {
+        rAmountsOut = ReservoirLibrary.getAmountsOut(address(factory), aAmountIn, aPath, aCurveIds);
+    }
+
+    function getAmountsIn(
+        uint256 aAmountOut,
+        address[] calldata aPath,
+        uint256[] calldata aCurveIds
+    ) external view returns(uint256[] memory rAmountsIn) {
+        rAmountsIn = ReservoirLibrary.getAmountsIn(address(factory), aAmountOut, aPath, aCurveIds);
+    }
+
+    function quoteAddLiquidity(
+        address aTokenA,
+        address aTokenB,
+        uint256 aCurveId,
+        uint256 aAmountADesired,
+        uint256 aAmountBDesired
+    ) external view returns (uint256 rAmountA, uint256 rAmountB, uint256 rLiq) {
+        address lPair = factory.getPair(aTokenA, aTokenB, aCurveId);
+        (uint lReserveA, uint lReserveB) = (0,0);
+        uint lTokenAPrecisionMultiplier = uint256(10) ** (18 - ERC20(aTokenA).decimals());
+        uint lTokenBPrecisionMultiplier = uint256(10) ** (18 - ERC20(aTokenB).decimals());
+        uint lTotalSupply = 0;
+
+        if (lPair != address(0)) {
+            lTotalSupply = IReservoirPair(lPair).totalSupply();
+            (lReserveA, lReserveB) = ReservoirLibrary.getReserves(address(factory), aTokenA, aTokenB, aCurveId);
+        }
+
+        if (lReserveA == 0 && lReserveB == 0) {
+            (rAmountA, rAmountB) = (aAmountADesired, aAmountBDesired);
+            if (aCurveId == 0) {
+                rLiq = FixedPointMathLib.sqrt(rAmountA * rAmountB) - MINIMUM_LIQUIDITY;
+            }
+            else if (aCurveId == 1) {
+                uint256 newLiq = ReservoirLibrary.computeStableLiquidity(
+                    rAmountA,
+                    rAmountB,
+                    lTokenAPrecisionMultiplier,
+                    lTokenBPrecisionMultiplier,
+                    2 * StablePair(lPair).getCurrentAPrecise()
```

Comment:
This 2 is the `N` in `N_A`?

Reply:
yes precisely since our implementation is only with 2 tokens.